File: Customer.cpp
///////////////////////////////////////////////////////////////////////////////
// Author: Katie Garner
// Description: This is the implementation file for the Customer class for the
//				COSC412 Project.
///////////////////////////////////////////////////////////////////////////////
#include "Customer.h"

// Constructors
Customer::Customer() {
	firstName = "";
	lastName = "";
	phoneNumber = "";
	address = "";
	email = "";
    salt = "";
	customerID = -1;
}
	
Customer::Customer(string fN, string lN, string pN, string add, string email, string uName, string pHash, string s, int custID) {
	this->firstName = fN;
	this->lastName = lN;
	this->phoneNumber = pN;
	this->address = add;
	this->email = email;
	this->userName = uName;
	this->passwordHash = pHash;
	this->salt = s;
	this->customerID = custID;
}

//Getters
string Customer::getFirstName() const
{
    return firstName;
}

string Customer::getLastName() const
{
    return lastName;
}

string Customer::getPhoneNumber() const
{
    return phoneNumber;
}

string Customer::getAddress() const
{
    return address;
}

string Customer::getEmail() const
{
    return email;
}

int Customer::getCustomerID() const
{
    return customerID;
}

// Setters
void Customer::setFirstName(const string& fN) {
    firstName = fN;
}

void Customer::setLastName(const string& lN) {
    lastName = lN;
}

void Customer::setPhoneNumber(const string& pN) {
    phoneNumber = pN;
}

void Customer::setAddress(const string& addr) {
    address = addr;
}

void Customer::setEmail(const string& em) {
    email = em;
}

void Customer::setUsername(const string& uName) {
    userName = uName;
}

void Customer::setPassword(const string& plainPassword) {
    passwordHash = HashUtilities::hashPassword(plainPassword, salt);
}

// Helper methods
int Customer::findCustomerID(const SearchCriteria& criteria) {
    // TODO: Replace with database query

    // The database will:
    // 1. Search the customers table
    // 2. Match fields based on non-empty criteria
    // 3. Return the matching customerID if found
    // 4. Return -1 if no match exists

    return -1;
}

int Customer::authenticateCustomer(const string& userName, const string& password) {
    // TODO: Replace with database query.

    // This method will:
    // 1. Search the customers table for the given username
    // 2. Retrieve the stored password hash and salt for that customer
    // 3. Hash the entered plainPassword using the stored salt
    // 4. Compare the new hash to the stored password hash
    // 5. Return the customerID if they match
    // 6. Return -1 if no matching username exists or the password is incorrect

    return -1;
}

bool Customer::createCustomerInDB() {
    // Validate required fields are provided
    if (firstName.empty() || lastName.empty() || email.empty() || userName.empty() || passwordHash.empty()) {
        return false;
    }
    
    // TODO: Replace with database insertion query.

    // This method will:
    // 1. Check whether the username or email already exists in the database
    // 2. Insert the new customer record into the customers table
    // 3. Store the customer's passwordHash and salt in the database
    // 4. Assign the generated customerID from the database to this object
    // 5. Return true if the insertion is successful
    // 6. Return false if creation fails or required fields are missing

    return true;
}

bool Customer::updateCustomerInDB() const {
    // TODO: Replace with database update query.
    
    // This method will:
    // 1. Validate that customerID is not -1
    // 2. Use the current object's customerID to locate the record in the database
    // 3. Update the customer's fields (first name, last name, email, etc.)
    // 4. Return true if the update is successful
    // 5. Return false if the customerID is invalid or the update fails

    return true;
}

bool Customer::deleteCustomerByID(int customerID) {
    // TODO: Replace with database deletion query.
    
    // This method will:
    // 1. Validate that customerID is not -1
    // 2. Use the provided customerID to locate the record in the database
    // 3. Delete the customer from the database
    // 4. Return true if the deletion is successful
    // 5. Return false if the customerID is invalid or the deletion fails

    return true;
}

File: customer.py
from hash_utilities import HashUtilities
from database import get_connection

# Customer class handles customer authentication and database operations
class Customer:
    """Initialize a Customer object with profile and login information"""
    def __init__(self, fN="", lN="", pN="", addr="", email="", custID=-1,
                 uName="", passwordHash="", salt="", mustChangePassword=False):
        self.firstName = fN
        self.lastName = lN
        self.phoneNumber = pN
        self.address = addr
        self.email = email
        self.customerID = custID
        self.userName = uName
        self.passwordHash = passwordHash
        self.salt = salt
        self.mustChangePassword = mustChangePassword
    
    def changePassword(self, newPassword):
        """Update customer's password (hash + salt) and clear must-change flag"""
        if not newPassword or self.customerID == -1:
            return False

        # Hash new password and update flags
        self.passwordHash, self.salt = HashUtilities.hash_password(newPassword)
        self.mustChangePassword = False

        conn = get_connection()
        cursor = conn.cursor()

        try:
            # Save updated credentials to database
            cursor.execute('''
                UPDATE customers
                SET passwordHash = ?, salt = ?, mustChangePassword = ?
                WHERE Customer_ID = ?
            ''', (self.passwordHash, self.salt, 0, self.customerID))

            conn.commit()

            if cursor.rowcount == 0:
                conn.close()
                return False

            conn.close()
            return True

        except:
            conn.close()
            return False

    @staticmethod
    def authenticateCustomer(userName, password):
        """Verify login credentials and return customer ID + password change status"""
        if not userName or not password:
            return None

        conn = get_connection()
        cursor = conn.cursor()

        # Fetch stored credentials for username
        cursor.execute('''
            SELECT Customer_ID, passwordHash, salt, mustChangePassword
            FROM customers
            WHERE username = ?
        ''', (userName,))

        row = cursor.fetchone()
        conn.close()

        if row is None:
            return None

        # Compare hashed password with stored hash
        entered_hash = HashUtilities.hash_with_salt(password, row["salt"])

        if entered_hash == row["passwordHash"]:
            return {
                "customerID": row["Customer_ID"],
                "mustChangePassword": bool(row["mustChangePassword"])
            }

        return None

    def createCustomerInDB(self):
        """Create a new customer in the database (generates username and temp password)"""
        if not self.firstName or not self.lastName or not self.email or not self.address or not self.phoneNumber:
            return False

        conn = get_connection()
        cursor = conn.cursor()

        try:
            # Check whether email already exists
            cursor.execute('''
                SELECT Customer_ID
                FROM customers
                WHERE email = ?
            ''', (self.email,))
            existing_email = cursor.fetchone()

            if existing_email is not None:
                conn.close()
                return False

            # Generate a base username from first initial + last name
            base_username = (self.firstName[0] + self.lastName).lower().replace(" ", "")
            username = base_username
            counter = 1

            # Make sure username is unique
            while True:
                cursor.execute('''
                    SELECT Customer_ID
                    FROM customers
                    WHERE username = ?
                ''', (username,))
                existing_username = cursor.fetchone()

                if existing_username is None:
                    break

                username = f"{base_username}{counter}"
                counter += 1

            # Generate temporary password and hash it
            temp_password = "temp123"
            self.passwordHash, self.salt = HashUtilities.hash_password(temp_password)
            self.userName = username
            self.mustChangePassword = True

            # Insert new customer record into database
            cursor.execute('''
                INSERT INTO customers
                (first_name, last_name, email, number, address, username, passwordHash, salt, mustChangePassword)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
            ''', (
                self.firstName,
                self.lastName,
                self.email,
                self.phoneNumber,
                self.address,
                self.userName,
                self.passwordHash,
                self.salt,
                1
            ))

            conn.commit()
            self.customerID = cursor.lastrowid
            conn.close()
            return True

        except:
            conn.close()
            return False

    def updateCustomerInDB(self):
        """Update existing customer information in the database"""
        if self.customerID == -1:
            return False

        if not self.firstName or not self.lastName or not self.email or not self.address or not self.phoneNumber:
            return False

        conn = get_connection()
        cursor = conn.cursor()

        try:
            # Check for duplicate email (excluding current customer)
            cursor.execute('''
                SELECT Customer_ID
                FROM customers
                WHERE email = ? AND Customer_ID != ?
            ''', (self.email, self.customerID))

            existing_email = cursor.fetchone()
            if existing_email is not None:
                conn.close()
                return False

            # Update with or without username depending on whether it exists
            if self.userName:
                cursor.execute('''
                    UPDATE customers
                    SET first_name = ?, last_name = ?, email = ?, number = ?, address = ?, username = ?
                    WHERE Customer_ID = ?
                ''', (
                    self.firstName,
                    self.lastName,
                    self.email,
                    self.phoneNumber,
                    self.address,
                    self.userName,
                    self.customerID
                ))
            else:
                cursor.execute('''
                    UPDATE customers
                    SET first_name = ?, last_name = ?, email = ?, number = ?, address = ?
                    WHERE Customer_ID = ?
                ''', (
                    self.firstName,
                    self.lastName,
                    self.email,
                    self.phoneNumber,
                    self.address,
                    self.customerID
                ))

            conn.commit()

            if cursor.rowcount == 0:
                conn.close()
                return False

            conn.close()
            return True

        except:
            conn.close()
            return False

    @staticmethod
    def findCustomerIDBySearch(searchText):
        """Find a customer ID using email, phone, full name, or username.
            Returns: ID, -1 (not found), or -2 (multiple matches)"""
        if not searchText:
            return -1

        searchText = searchText.strip()
        conn = get_connection()
        cursor = conn.cursor()

        try:
            # Try matching by email first
            if "@" in searchText:
                cursor.execute('''
                    SELECT Customer_ID
                    FROM customers
                    WHERE LOWER(email) = LOWER(?)
                ''', (searchText,))
                row = cursor.fetchone()
                conn.close()
                return row["Customer_ID"] if row else -1

            # Normalize input to digits and compare phone numbers
            digits_only = "".join(char for char in searchText if char.isdigit())
            if len(digits_only) >= 7:
                cursor.execute('''
                    SELECT Customer_ID, number
                    FROM customers
                ''')
                rows = cursor.fetchall()

                for row in rows:
                    db_digits = "".join(char for char in row["number"] if char.isdigit()) if row["number"] else ""
                    if db_digits == digits_only:
                        conn.close()
                        return row["Customer_ID"]

                conn.close()
                return -1

            # Attempt match using first and last name
            if " " in searchText:
                parts = searchText.split()
                if len(parts) >= 2:
                    first_name = parts[0]
                    last_name = " ".join(parts[1:])

                    cursor.execute('''
                        SELECT Customer_ID
                        FROM customers
                        WHERE LOWER(first_name) = LOWER(?) AND LOWER(last_name) = LOWER(?)
                    ''', (first_name, last_name))

                    rows = cursor.fetchall()

                    if len(rows) == 1:
                        conn.close()
                        return rows[0]["Customer_ID"]

                    if len(rows) > 1:
                        conn.close()
                        return -2

            # Fallback: try matching by username
            cursor.execute('''
                SELECT Customer_ID
                FROM customers
                WHERE LOWER(username) = LOWER(?)
            ''', (searchText,))
            row = cursor.fetchone()
            conn.close()

            return row["Customer_ID"] if row else -1

        except:
            conn.close()
            return -1

    def deleteCustomerInDB(self):
        """Delete customer from the database using their ID"""
        if self.customerID == -1:
            return False

        conn = get_connection()
        cursor = conn.cursor()

        try:
            # Remove customer record from database
            cursor.execute('''
                DELETE FROM customers
                WHERE Customer_ID = ?
            ''', (self.customerID,))

            conn.commit()

            if cursor.rowcount == 0:
                conn.close()
                return False

            self.customerID = -1
            conn.close()
            return True

        except:
            conn.close()
            return False

    @staticmethod
    def getCustomerInfoByID(customerID):
        """Retrieve customer profile information by ID for display purposes"""
        if customerID == -1:
            return None

        conn = get_connection()
        cursor = conn.cursor()

        try:
            # Query database for customer details
            cursor.execute('''
                SELECT Customer_ID, first_name, last_name, email, number, address, username
                FROM customers
                WHERE Customer_ID = ?
            ''', (customerID,))

            row = cursor.fetchone()
            conn.close()

            if row is None:
                return None

            # Convert database row into dictionary for frontend use
            return {
                "customerID": row["Customer_ID"],
                "firstName": row["first_name"],
                "lastName": row["last_name"],
                "email": row["email"],
                "phoneNumber": row["number"],
                "address": row["address"],
                "username": row["username"]
            }

        except:
            conn.close()
            return None